### PR TITLE
Python3 compatibility

### LINKED ISSUE #1039: Does gpkit work with python3?
Hi,

When I tried to setup gpkit with python3, it didnt work out.

Particularly, the commands "pip install ctypesgen --pre" and "pip install gpkit" does not work for my Ubuntu 16.04 pc. You can find the error messages in below.

Thanks a lot!

```
Error message for "pip install ctypesgen --pre"
Collecting ctypesgen
  Using cached ctypesgen-0.r125.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-0teb07md/ctypesgen/setup.py", line 13, in <module>
        import ctypesgencore
      File "/tmp/pip-build-0teb07md/ctypesgen/ctypesgencore/__init__.py", line 56, in <module>
        import processor
    ImportError: No module named 'processor'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-0teb07md/ctypesgen/

Error message for "pip install gpkit"



Collecting gpkit
  Using cached gpkit-0.5.2.0.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-5gnqcug8/gpkit/setup.py", line 8, in <module>
        from gpkit.build import build_gpkit
      File "/tmp/pip-build-5gnqcug8/gpkit/gpkit/__init__.py", line 117, in <module>
        enable_units()
      File "/tmp/pip-build-5gnqcug8/gpkit/gpkit/__init__.py", line 71, in enable_units
        units = GPkitUnits()
      File "/tmp/pip-build-5gnqcug8/gpkit/gpkit/__init__.py", line 36, in __init__
        if hasattr(ureg, "__nonzero__"):
      File "/home/lmc3696/myvirtualenv/lib/python3.5/site-packages/pint/unit.py", line 375, in __getattr__
        return self.Unit(item)
      File "/home/lmc3696/myvirtualenv/lib/python3.5/site-packages/pint/unit.py", line 66, in __new__
        inst._units = inst._REGISTRY.parse_units(units)._units
      File "/home/lmc3696/myvirtualenv/lib/python3.5/site-packages/pint/unit.py", line 1201, in parse_units
        units = self._parse_units(input_string, as_delta)
      File "/home/lmc3696/myvirtualenv/lib/python3.5/site-packages/pint/unit.py", line 1223, in _parse_units
        cname = self.get_name(name)
      File "/home/lmc3696/myvirtualenv/lib/python3.5/site-packages/pint/unit.py", line 753, in get_name
        raise UndefinedUnitError(name_or_alias)
    pint.errors.UndefinedUnitError: '__nonzero__' is not defined in the unit registry
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-5gnqcug8/gpkit/
```

### PR DESCRIPTION
**Pull request by whoburg**
closes #1039. Might be good to rebase once #1121 is merged.

@galbramc, is there a smart way to add this functionality to our tests?

### CODE REVIEW

**Comment by bqpd** [comment]
test this please

**Comment by bqpd** [comment]
@whoburg we could easily add python3 unit tests to jenkins, but still don't pass all tests on python3; this PR just makes it so that at least some of the examples can run

**Comment by galbramc** [comment]
when python3 works we can add more tests that use python3

**Comment by whoburg** [comment]
then this doesn't really close #1039...

**Comment by whoburg** [comment]
just rebased to fix a merge conflict and force-pushed -- please update local copies of this branch

File: gpkit/nomials/nomial_math.py [hunk 1]
```diff
@@ -346,6 +348,8 @@ def __le__(self, other):
 
     # Posynomial.__ge__ falls back on Signomial.__ge__
 
+    __hash__ = Signomial.__hash__  # required by Python 3
+
```
Comment (whoburg) [inline gpkit/nomials/nomial_math.py]: interesting. Specifically,

"If a class that overrides __eq__() needs to retain the implementation of __hash__() from a parent class, the interpreter must be told this explicitly by setting __hash__ = <ParentClass>.__hash__."  (https://docs.python.org/3/reference/datamodel.html#object.__hash__)

File: gpkit/small_classes.py [hunk 2]
```diff
@@ -35,13 +33,17 @@ def to_(self):  # used in tocoo, tocsc, etc below
...
-    def __init__(self, *args, **kwargs):
-        super(CootMatrix, self).__init__(*args, **kwargs)
+    def __init__(self, row, col, data):
+        self.row, self.col, self.data = row, col, data
         self.shape = [(max(self.row) + 1) if self.row else 0,
                       (max(self.col) + 1) if self.col else 0]
 
+    def __eq__(self, other):
+        return (self.row == other.row and self.col == other.col
+                and self.data == other.data and self.shape == other.shape)
```
Comment (whoburg) [inline gpkit/small_classes.py]: would it be marginally faster to compare shape first (fail fast)? Or are the vast majority of comparisons expected to be between objects with the same shape?

**Comment by whoburg** [comment]
I created this PR from @bqpd's commits, so I can't approve it, but it's reviewed and approved by me. Todo before merging:

 - [ ] Need model tests to pass
 - [ ] Make sure #1039 is still open (I don't consider that closed until we test Python3 on Jenkins)

**Comment by whoburg** [comment]
test models please

**Comment by galbramc** [comment]
You need to decide when you want python3 on jenkins. As soon as I add python3 to jeninks, all other testing will fail. I can't add python3 for just this PR, it will be all PRs.

**Comment by whoburg** [comment]
understood, we'll let you know. When you say "all other testing will fail", you mean it will fail because our tests don't currently all pass on Python3?

Is it possible to test both Python 2 and Python 3 in the future when we support both?

**Comment by galbramc** [comment]
Yes we can test with both Python 2 and Python 3. However, as soon as I add the Python 3 test, all other PRs will fail those tests. So if you want me to add Python 3 tests before this PR is merged, then you have to be ready to stop working on the other PRs while we get the Python 3 testing setup and this PR merged. Does that make sense?

**Pull request closed without merging**